File: src/System.Windows.Forms.Analyzers.VisualBasic/tests/UnitTests/System.Windows.Forms.Analyzers.VisualBasic.Tests/Analyzers/Designer/DesignerAnalyzerTests.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.

Imports System.Windows.Forms.Analyzers.Tests
Imports System.Windows.Forms.VisualBasic.Analyzers.Designer
Imports Microsoft.CodeAnalysis.Testing
Imports Xunit

''' <summary>
'''  Tests the Visual Basic WinForms Designer guardrails.
''' </summary>
<ForceGC()>
<SkipOnArchitecture(TestArchitectures.X86, "Analyzer tests require Roslyn reference assembly extraction")>
Public Class DesignerAnalyzerTests

    Private Const MainSource As String = "
Imports System.Windows.Forms

Namespace Test

    Partial Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub
    End Class
End Namespace
"

    <Fact>
    Public Async Function InitializeComponent_ValidGeneratedCode_NoDiagnostic() As Task
        Const designerSource As String = "
Imports System.ComponentModel
Imports System.Windows.Forms

Namespace Test

    Partial Class Form1
        Private components As IContainer

        Protected Overrides Sub Dispose(disposing As Boolean)
            components?.Dispose()
            MyBase.Dispose(disposing)
        End Sub

        Private Sub InitializeComponent()
            Button1 = New Button()
            Controls.Add(Button1)
        End Sub

        Private WithEvents Button1 As Button
    End Class
End Namespace
"

        Dim testCase As AnalyzerTestCase = CreateTestCase(designerSource)

        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of InitializeComponentAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)
        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of DesignerFileStructureAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)
    End Function

    <Fact>
    Public Async Function InitializeComponent_ControlFlow_ReportsDiagnostics() As Task
        Const designerSource As String = "
Imports System.Windows.Forms

Namespace Test

    Partial Class Form1
        Private Sub InitializeComponent()
            {|WFO3000:For|} i As Integer = 0 To Controls.Count
            Next

            {|WFO3000:For|} Each control As Control In Controls
            Next

            {|WFO3000:While|} DesignMode
            End While

            {|WFO3000:Do|}
            Loop While DesignMode

            {|WFO3000:If|} DesignMode Then
            End If

            {|WFO3000:If|} DesignMode Then Text = Name

            {|WFO3000:Select|} Case Controls.Count
            End Select

            {|WFO3000:GoTo|} Done
Done:
            {|WFO3000:Try|}
            Catch
            End Try

            {|WFO3000:SyncLock|} Me
            End SyncLock
        End Sub
    End Class
End Namespace
"

        Dim testCase As AnalyzerTestCase = CreateTestCase(designerSource)

        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of InitializeComponentAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)
    End Function

    <Fact>
    Public Async Function InitializeComponent_CodeDomUnsupportedExpressions_ReportDiagnostics() As Task
        Const designerSource As String = "
Imports System
Imports System.Windows.Forms

Namespace Test

    Partial Class Form1
        Private Sub InitializeComponent()
            Dim name As String = {|WFO3005:NameOf|}(Form1)
            Dim text As String = {|WFO3009:$""|}{name}: {Controls.Count}""
            Dim handler As EventHandler = {|WFO3010:Sub|}(sender, e) Text = text
            Dim value As Object = {|WFO3007:If|}(Tag, Me)
            Dim control As Control = {|WFO3008:Parent|}?.Parent
            Dim count As Integer = {|WFO3006:If|}(DesignMode, 0, Controls.Count)
        End Sub
    End Class
End Namespace
"

        Dim testCase As AnalyzerTestCase = CreateTestCase(designerSource)

        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of InitializeComponentAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)
    End Function

    <Fact>
    Public Async Function DesignerMembers_ReportStructuralDiagnostics() As Task
        Const designerSource As String = "
Imports System
Imports System.Windows.Forms

Namespace Test

    Partial Class Form1
        Private WithEvents Button1 As Button

        Private Sub InitializeComponent()
            Button1 = New Button()
        End Sub

        Public Property {|WFO3001:Value|} As Integer
        Public Event {|WFO3003:Changed|} As EventHandler
        Public Delegate Sub {|WFO3003:Callback|}()
    End Class
End Namespace
"

        Dim testCase As AnalyzerTestCase = CreateTestCase(designerSource)
        testCase.ExpectedDiagnostics.Add(
            New DiagnosticResult("WFO3002", Microsoft.CodeAnalysis.DiagnosticSeverity.Warning).
                WithSpan("/0/Form1.Designer.vb", 8, 28, 8, 35).
                WithArguments("Button1"))

        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of DesignerFileStructureAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)
    End Function

    <Fact>
    Public Async Function MainFileField_ReferencedByInitializeComponent_ReportsDiagnostic() As Task
        Const mainSource As String = "
Imports System.Windows.Forms

Namespace Test

    Partial Class Form1
        Inherits Form

        Private WithEvents {|WFO3002:Button1|} As Button

        Public Sub New()
            InitializeComponent()
        End Sub
    End Class
End Namespace
"

        Const designerSource As String = "
Imports System.Windows.Forms

Namespace Test

    Partial Class Form1
        Private Sub InitializeComponent()
            Button1 = New Button()
            Controls.Add(Button1)
        End Sub
    End Class
End Namespace
"

        Dim testCase As New AnalyzerTestCase(
            ReferenceAssemblies.Net.Net90Windows,
            New AnalyzerTestSource("Form1.vb", mainSource),
            New AnalyzerTestSource("Form1.Designer.vb", designerSource))

        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of DesignerFileStructureAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)
    End Function

    <Fact>
    Public Async Function DesignerFile_MalformedMembers_ReportUnexpectedMember() As Task
        Const designerSource As String = "
Namespace Test

    Partial Class Form1
        Private Sub InitializeComponent()
        End Sub

        Private Sub {|WFO3001:Dispose|}(disposing As Integer)
        End Sub
    End Class
End Namespace
"

        Dim testCase As AnalyzerTestCase = CreateTestCase(designerSource)

        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of DesignerFileStructureAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)
    End Function

    <Fact>
    Public Async Function UnrelatedDesignerNamedFile_NoDiagnostic() As Task
        Const mainSource As String = "
Namespace Test

    Partial Class Model
    End Class
End Namespace
"

        Const designerSource As String = "
Namespace Test

    Partial Class Model
        Private Sub InitializeComponent()
            If True Then
            End If
        End Sub

        Public Property Value As Integer
    End Class
End Namespace
"

        Dim testCase As New AnalyzerTestCase(
            ReferenceAssemblies.Net.Net90Windows,
            New AnalyzerTestSource("Model.vb", mainSource),
            New AnalyzerTestSource("Model.Designer.vb", designerSource))

        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of InitializeComponentAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)
        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of DesignerFileStructureAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)
    End Function

    <Fact>
    Public Async Function InitializeComponent_DiagnosticSuppressedByAnalyzerConfig() As Task
        Const designerSource As String = "
Namespace Test

    Partial Class Form1
        Private Sub InitializeComponent()
            If DesignMode Then
            End If
        End Sub
    End Class
End Namespace
"

        Dim testCase As AnalyzerTestCase = CreateTestCase(designerSource)
        testCase.AnalyzerConfigFiles.Add(
            (
                "/.globalconfig",
                "
is_global = true

dotnet_diagnostic.WFO3000.severity = none
"))

        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of InitializeComponentAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)
    End Function

    Private Shared Function CreateTestCase(designerSource As String) As AnalyzerTestCase
        Return New AnalyzerTestCase(
            ReferenceAssemblies.Net.Net90Windows,
            New AnalyzerTestSource("Form1.vb", MainSource),
            New AnalyzerTestSource("Form1.Designer.vb", designerSource))
    End Function

    <Theory>
    <InlineData("InitializeComponent")>
    <InlineData("initializecomponent")>
    <InlineData("INITIALIZECOMPONENT()")>
    Public Async Function InitializeComponent_OptionalSyntax_ReportsControlFlow(declaration As String) As Task
        Dim testCase As AnalyzerTestCase = CreateTestCase(
            $"
Namespace Test
    Partial Class Form1
        Private Sub {declaration}
            {{|WFO3000:If|}} True Then
            End If
        End Sub
    End Class
End Namespace
")

        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of InitializeComponentAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)

        Dim structureCase As AnalyzerTestCase = CreateTestCase(
            testCase.Sources(1).Source.Replace("{|WFO3000:If|}", "If"))
        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(
            Of DesignerFileStructureAnalyzer)(structureCase).
            RunAsync(TestContext.Current.CancellationToken)
    End Function

    <Theory>
    <InlineData("System.ComponentModel.Component", True)>
    <InlineData("ActualComponent", True)>
    <InlineData("DerivedComponent", True)>
    <InlineData("Other.Component", False)>
    <InlineData("Other.Control", False)>
    Public Async Function DesignerType_UsesFrameworkIdentity(baseType As String, expected As Boolean) As Task
        Dim condition As String = If(expected, "{|WFO3000:If|}", "If")
        Dim testCase As New AnalyzerTestCase(
            ReferenceAssemblies.Net.Net90Windows,
            New AnalyzerTestSource("Form1.vb", $"
Imports ActualComponent = System.ComponentModel.Component
Class DerivedComponent
    Inherits ActualComponent
End Class
Namespace Other
    Class Component
    End Class
    Class Control
    End Class
End Namespace
Partial Class Form1
    Inherits {baseType}
End Class
"),
            New AnalyzerTestSource("Form1.Designer.vb", $"
Partial Class Form1
    Private Sub InitializeComponent()
        {condition} True Then
        End If
    End Sub
End Class
"))
        Dim test = AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(Of InitializeComponentAnalyzer)(testCase)
        test.SolutionTransforms.Add(
            Function(solution, projectId) solution.WithProjectCompilationOptions(
                projectId,
                DirectCast(solution.GetProject(projectId).CompilationOptions,
                    Microsoft.CodeAnalysis.VisualBasic.VisualBasicCompilationOptions).WithRootNamespace("TestRoot")))

        Await test.RunAsync(TestContext.Current.CancellationToken)
    End Function

    <Fact>
    Public Async Function InitializeComponent_UsingAndAwait_ReportDiagnostics() As Task
        Dim testCase As AnalyzerTestCase = CreateTestCase(
            "
Namespace Test
    Partial Class Form1
        Private Async Sub InitializeComponent()
            {|WFO3000:Using|} component As New System.ComponentModel.Component()
            End Using
            {|WFO3000:Await|} System.Threading.Tasks.Task.CompletedTask
        End Sub
    End Class
End Namespace
")
        Await AnalyzerTestFactory.CreateVisualBasicAnalyzerTest(Of InitializeComponentAnalyzer)(testCase).
            RunAsync(TestContext.Current.CancellationToken)
    End Function
End Class

File: src/System.Windows.Forms.Analyzers.CSharp/tests/UnitTests/Analyzers/Designer/DesignerAnalyzerTests.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

using System.Windows.Forms.CSharp.Analyzers.Designer;
using Microsoft.CodeAnalysis.Testing;

namespace System.Windows.Forms.Analyzers.Tests;

/// <summary>
///  Tests the C# WinForms Designer guardrails.
/// </summary>
[ForceGC]
[SkipOnArchitecture(TestArchitectures.X86, "Analyzer tests require Roslyn reference assembly extraction")]
public class DesignerAnalyzerTests
{
    private const string MainSource =
        """
        using System.Windows.Forms;

        namespace Test;

        partial class Form1 : Form
        {
            public Form1()
            {
                InitializeComponent();
            }
        }
        """;

    [Fact]
    public async Task InitializeComponent_ValidGeneratedCode_NoDiagnostic()
    {
        const string designerSource =
            """
            using System.ComponentModel;
            using System.Windows.Forms;

            namespace Test;

            partial class Form1
            {
                private IContainer? components;

                protected override void Dispose(bool disposing)
                {
                    components?.Dispose();
                    base.Dispose(disposing);
                }

                private void InitializeComponent()
                {
                    button1 = new Button();
                    Controls.Add(button1);
                }

                private Button button1;
            }
            """;

        AnalyzerTestCase testCase = CreateTestCase(designerSource);

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<DesignerFileStructureAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task InitializeComponent_ControlFlow_ReportsDiagnostic()
    {
        const string designerSource =
            """
            using System.Windows.Forms;

            namespace Test;

            partial class Form1
            {
                private void InitializeComponent()
                {
                    {|WFO3000:if|} (true)
                    {
                        Controls.Add(new Button());
                    }
                }
            }
            """;

        AnalyzerTestCase testCase = CreateTestCase(designerSource);

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Theory]
    [InlineData("{|WFO3000:for|} (int i = 0; i < Controls.Count; i++) { }")]
    [InlineData("{|WFO3000:foreach|} (Control control in Controls) { }")]
    [InlineData("{|WFO3000:while|} (DesignMode) { }")]
    [InlineData("{|WFO3000:do|} { } while (DesignMode);")]
    [InlineData("{|WFO3000:if|} (DesignMode) { }")]
    [InlineData("{|WFO3000:switch|} (Controls.Count) { default: break; }")]
    [InlineData("_ = Controls.Count {|WFO3000:switch|} { _ => 0 };")]
    [InlineData("void {|WFO3000:Configure|}() { }")]
    [InlineData("{|WFO3000:goto|} End; End:;")]
    [InlineData("{|WFO3000:try|} { } catch { }")]
    [InlineData("{|WFO3000:lock|} (this) { }")]
    [InlineData("Tag {|WFO3007:??=|} new object();")]
    [InlineData("{|WFO3000:foreach|} (var (x, y) in new (int, int)[] { (1, 2) }) { }")]
    [InlineData("{|WFO3000:using|} (var component = new System.ComponentModel.Component()) { }")]
    [InlineData("{|WFO3000:using|} var component = new System.ComponentModel.Component();")]
    public async Task InitializeComponent_UnsupportedConstruct_ReportsDiagnostic(string statement)
    {
        string designerSource =
            $$"""
            using System.Windows.Forms;

            namespace Test;

            partial class Form1
            {
                private void InitializeComponent()
                {
                    {{statement}}
                }
            }
            """;

        AnalyzerTestCase testCase = CreateTestCase(designerSource);

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task InitializeComponent_CodeDomUnsupportedExpressions_ReportDiagnostics()
    {
        const string designerSource =
            """
            using System;
            using System.Windows.Forms;

            namespace Test;

            partial class Form1
            {
                private void InitializeComponent()
                {
                    string name = {|WFO3005:nameof|}(Form1);
                    string text = {|WFO3009:$"|}{name}: {Controls.Count}";
                    EventHandler handler = (_, _) {|WFO3010:=>|} Text = text;
                    EventHandler alternate = {|WFO3010:delegate|} { };
                    object value = Tag {|WFO3007:??|} this;
                    Control? control = Parent{|WFO3008:?|}.Parent;
                    int count = DesignMode {|WFO3006:?|} 0 : Controls.Count;
                }
            }
            """;

        AnalyzerTestCase testCase = CreateTestCase(designerSource);

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task DesignerFile_CollectionExpression_ReportsDiagnostic()
    {
        const string designerSource =
            """
            using System.Windows.Forms;

            namespace Test;

            partial class Form1
            {
                private void InitializeComponent()
                {
                    Control[] controls = {|WFO3004:[|}new Button()];
                }
            }
            """;

        AnalyzerTestCase testCase = CreateTestCase(designerSource);

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<DesignerFileStructureAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task DesignerMembers_ReportStructuralDiagnostics()
    {
        const string designerSource =
            """
            using System;
            using System.Windows.Forms;

            namespace Test;

            partial class Form1
            {
                private Button {|WFO3002:button1|};

                private void InitializeComponent()
                {
                    button1 = new Button();
                }

                public int {|WFO3001:Value|} { get; set; }
                public event EventHandler? {|WFO3003:Changed|};
                public delegate void {|WFO3003:Callback|}();
            }
            """;

        AnalyzerTestCase testCase = CreateTestCase(designerSource);

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<DesignerFileStructureAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task MainFileField_ReferencedByInitializeComponent_ReportsDiagnosticAtDeclaration()
    {
        const string mainSource =
            """
            using System.Windows.Forms;

            namespace Test;

            partial class Form1 : Form
            {
                private Button {|WFO3002:button1|};

                public Form1()
                {
                    InitializeComponent();
                }
            }
            """;

        const string designerSource =
            """
            using System.Windows.Forms;

            namespace Test;

            partial class Form1
            {
                private void InitializeComponent()
                {
                    button1 = new Button();
                    Controls.Add(button1);
                }
            }
            """;

        AnalyzerTestCase testCase = new(
            ReferenceAssemblies.Net.Net90Windows,
            new AnalyzerTestSource("Form1.cs", mainSource),
            new AnalyzerTestSource("Form1.Designer.cs", designerSource));

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<DesignerFileStructureAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task InitializeComponent_LocalAndMethodReferences_DoNotReportFieldDiagnostic()
    {
        const string designerSource =
            """
            using System.Windows.Forms;

            namespace Test;

            partial class Form1
            {
                private void InitializeComponent()
                {
                    Button button1 = CreateButton();
                    Controls.Add(button1);
                }

                private Button {|WFO3001:CreateButton|}()
                {
                    return new Button();
                }
            }
            """;

        AnalyzerTestCase testCase = CreateTestCase(designerSource);

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<DesignerFileStructureAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task DesignerFile_MultipleTypes_AnalyzesMatchingControlOnly()
    {
        const string mainSource =
            """
            using System.Windows.Forms;

            namespace Test;

            partial class Form1<T> : Form
            {
                public Form1()
                {
                    InitializeComponent();
                }

                public partial class Nested : UserControl
                {
                }
            }

            partial class Model
            {
            }
            """;

        const string designerSource =
            """
            namespace Test;

            partial class Form1<T>
            {
                private void InitializeComponent()
                {
                    {|WFO3000:if|} (DesignMode) { }
                }

                public partial class Nested
                {
                    private void InitializeComponent()
                    {
                    }

                    public int Value { get; set; }
                }
            }

            partial class Model
            {
                private void InitializeComponent()
                {
                    if (true) { }
                }

                public int Value { get; set; }
            }
            """;

        AnalyzerTestCase testCase = new(
            ReferenceAssemblies.Net.Net90Windows,
            new AnalyzerTestSource("Form1.cs", mainSource),
            new AnalyzerTestSource("Form1.Designer.cs", designerSource));

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task DesignerFile_MalformedDispose_ReportsUnexpectedMember()
    {
        const string designerSource =
            """
            namespace Test;

            partial class Form1
            {
                private void InitializeComponent()
                {
                }

                private void {|WFO3001:Dispose|}(int disposing)
                {
                }
            }
            """;

        AnalyzerTestCase testCase = CreateTestCase(designerSource);

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<DesignerFileStructureAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task DesignerFile_MalformedInitializeComponent_IsNotDesignerPair()
    {
        const string designerSource =
            """
            namespace Test;

            partial class Form1
            {
                private int InitializeComponent()
                {
                    if (DesignMode) { }
                    return 0;
                }

                public int Value { get; set; }
            }
            """;

        AnalyzerTestCase testCase = CreateTestCase(designerSource);

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<DesignerFileStructureAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task InitializeComponent_DiagnosticSuppressedByAnalyzerConfig()
    {
        const string designerSource =
            """
            namespace Test;

            partial class Form1
            {
                private void InitializeComponent()
                {
                    if (DesignMode) { }
                }
            }
            """;

        AnalyzerTestCase testCase = CreateTestCase(designerSource);
        testCase.AnalyzerConfigFiles.Add(
            ("/.globalconfig",
            """
            is_global = true

            dotnet_diagnostic.WFO3000.severity = none
            """));

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task UnrelatedDesignerNamedFile_NoDiagnostic()
    {
        const string mainSource =
            """
            namespace Test;

            partial class Model
            {
            }
            """;

        const string designerSource =
            """
            namespace Test;

            partial class Model
            {
                private void InitializeComponent()
                {
                    if (true)
                    {
                    }
                }

                public int Value { get; set; }
            }
            """;

        AnalyzerTestCase testCase = new(
            ReferenceAssemblies.Net.Net90Windows,
            new AnalyzerTestSource("Model.cs", mainSource),
            new AnalyzerTestSource("Model.Designer.cs", designerSource));

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<DesignerFileStructureAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    private static AnalyzerTestCase CreateTestCase(string designerSource)
        => new(
            ReferenceAssemblies.Net.Net90Windows,
            new AnalyzerTestSource("Form1.cs", MainSource),
            new AnalyzerTestSource("Form1.Designer.cs", designerSource));

    [Fact]
    public async Task InitializeComponent_ExpressionBody_ReportsDiagnostics()
    {
        AnalyzerTestCase testCase = CreateTestCase(
            """
            namespace Test;
            partial class Form1
            {
                private void InitializeComponent() {|WFO3000:=>|} Tag = Tag {|WFO3007:??|} new object();
            }
            """);

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task InitializeComponent_NonDesignerOwnedFields_NoDiagnostic()
    {
        AnalyzerTestCase testCase = new(
            ReferenceAssemblies.Net.Net90Windows,
            new AnalyzerTestSource("Form1.cs",
                """
                using System.Windows.Forms;
                namespace Test;
                class BaseForm : Form { protected Button inherited; }
                class External { public static Button Value; }
                partial class Form1 : BaseForm
                {
                    private readonly int customValue = 42;
                    private Button runtimeButton = new Button();
                }
                """),
            new AnalyzerTestSource("Form1.Designer.cs",
                """
                using System.Windows.Forms;
                namespace Test;
                partial class Form1
                {
                    private void InitializeComponent()
                    {
                        inherited = new Button();
                        External.Value = new Button();
                        TabIndex = customValue;
                        Controls.Add(runtimeButton);
                    }
                }
                """));

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<DesignerFileStructureAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Theory]
    [InlineData("System.ComponentModel.Component", true)]
    [InlineData("ActualComponent", true)]
    [InlineData("DerivedComponent", true)]
    [InlineData("System.Windows.Forms.UserControl", true)]
    [InlineData("Other.Component", false)]
    [InlineData("Other.Control", false)]
    public async Task DesignerType_UsesFrameworkIdentity(string baseType, bool expected)
    {
        string condition = expected ? "{|WFO3000:if|}" : "if";
        AnalyzerTestCase testCase = new(
            ReferenceAssemblies.Net.Net90Windows,
            new AnalyzerTestSource("Form1.cs",
                $$"""
                using ActualComponent = System.ComponentModel.Component;
                class DerivedComponent : ActualComponent { }
                namespace Other { class Component { } class Control { } }
                namespace Test { partial class Form1 : {{baseType}} { } }
                """),
            new AnalyzerTestSource("Form1.Designer.cs",
                $$"""
                namespace Test
                {
                    partial class Form1
                    {
                        private void InitializeComponent()
                        {
                            {{condition}} (true) { }
                        }
                    }
                }
                """));

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task InitializeComponent_MethodNamedNameof_NoDiagnostic()
    {
        AnalyzerTestCase testCase = CreateTestCase(
            """
            namespace Test;
            partial class Form1
            {
                private void InitializeComponent() { Text = nameof(1); }
            }
            """);
        testCase.Sources.Add(new AnalyzerTestSource("Methods.cs",
            """
            namespace Test;
            partial class Form1 { private string nameof(int value) => value.ToString(); }
            """));

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task InitializeComponent_AsyncAwait_ReportsDiagnostic()
    {
        AnalyzerTestCase testCase = CreateTestCase(
            """
            namespace Test;
            partial class Form1
            {
                private async void InitializeComponent()
                {
                    {|WFO3000:await|} System.Threading.Tasks.Task.CompletedTask;
                }
            }
            """);

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<InitializeComponentAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }

    [Fact]
    public async Task DesignerFile_NestedCollectionExpression_ReportsOnce()
    {
        AnalyzerTestCase testCase = CreateTestCase(
            """
            namespace Test;
            partial class Form1
            {
                private void InitializeComponent() { }
                partial class {|WFO3001:Nested|}
                {
                    private void InitializeComponent() { int[] values = {|WFO3004:[|}1, 2]; }
                }
            }
            """);
        testCase.Sources.Add(new AnalyzerTestSource("Nested.cs",
            """
            namespace Test;
            partial class Form1
            {
                partial class Nested : System.Windows.Forms.Form { }
            }
            """));

        await AnalyzerTestFactory.CreateCSharpAnalyzerTest<DesignerFileStructureAnalyzer>(testCase)
            .RunAsync(TestContext.Current.CancellationToken);
    }
}
